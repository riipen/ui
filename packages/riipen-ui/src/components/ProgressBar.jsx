import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class ProgressBar extends React.Component {
  static propTypes = {
    /**
     * List of additional classes to apply to this component.
     */
    classes: PropTypes.array,

    /**
     * The color to use.
     */
    color: PropTypes.oneOf([
      "negative",
      "positive",
      "primary",
      "secondary",
      "tertiary",
      "warning"
    ]),

    unfilled: PropTypes.oneOf(["light", "dark"]),

    /**
     * The percentage of the completion between 0 and 1.
     */
    completion: PropTypes.number
  };

  static defaultProps = {
    classes: [],
    color: "primary",
    completion: 0,
    unfilled: "light"
  };

  static contextType = ThemeContext;

  render() {
    const { classes, color, completion, unfilled } = this.props;

    const theme = this.context;

    const className = clsx(color, unfilled, classes);

    return (
      <div>
        <progress className={className} value={completion} />
        <style jsx>{`
          progress {
            -moz-appearance: none;
            -webkit-appearance: none;
            border: none;
            border-radius: ${theme.shape.borderRadius.md};
            height: 8px;
            width: 100%;
          }

          progress::-webkit-progress-bar {
            border-radius: ${theme.shape.borderRadius.md};
          }

          progress::-webkit-progress-value {
            border-radius: ${theme.shape.borderRadius.md};
          }

          progress::-moz-progress-bar {
            border-radius: ${theme.shape.borderRadius.md};
          }

          .primary::-webkit-progress-value {
            background: ${theme.palette.primary.main};
          }

          .primary::-moz-progress-bar {
            background: ${theme.palette.primary.main};
          }

          .light {
            background: ${theme.palette.grey[50]};
          }

          .light::-webkit-progress-bar {
            background: ${theme.palette.grey[50]};
          }

          .dark {
            background: ${theme.palette.grey[300]};
          }

          .dark::-webkit-progress-bar {
            background: ${theme.palette.grey[300]};
          }

          .secondary::-webkit-progress-value {
            background: ${theme.palette.secondary.main};
          }

          .secondary::-moz-progress-bar {
            background: ${theme.palette.secondary.main};
          }

          .tertiary::-webkit-progress-value {
            background: ${theme.palette.tertiary.main};
          }

          .tertiary::-moz-progress-bar {
            background: ${theme.palette.tertiary.main};
          }

          .positive::-webkit-progress-value {
            background: ${theme.palette.positive.main};
          }

          .positive::-moz-progress-bar {
            background: ${theme.palette.positive.main};
          }

          .negative::-webkit-progress-value {
            background: ${theme.palette.negative.main};
          }

          .negative::-moz-progress-bar {
            background: ${theme.palette.negative.main};
          }

          .warning::-webkit-progress-value {
            background: ${theme.palette.warning.main};
          }

          .warning::-moz-progress-bar {
            background: ${theme.palette.warning.main};
          }
        `}</style>
      </div>
    );
  }
}

export default withClasses()(ProgressBar);
