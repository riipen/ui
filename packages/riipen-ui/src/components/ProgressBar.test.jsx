import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import ProgressBar from "./ProgressBar.jsx";

describe("<ProgressBar>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<ProgressBar />);
    } catch (e) {
      error = e;
    }
    expect(error).toEqual(undefined);
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new ProgressBar().type.defaultProps;

      const wrapper = mount(<ProgressBar />);

      const component = wrapper.find("ProgressBar");
      expect(component.props().color).toEqual(defaultProps.color);
      expect(component.props().completion).toEqual(defaultProps.completion);
      expect(component.props().unfilled).toEqual(defaultProps.unfilled);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the progress element", () => {
      const classes = ["classOne"];

      const wrapper = mount(<ProgressBar classes={classes} />);

      expect(
        wrapper
          .find("ProgressBar")
          .find("progress")
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-progressbar"];

      const wrapper = mount(<ProgressBar />);

      expect(
        wrapper
          .find("ProgressBar")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<ProgressBar classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("applies valid custom colour class to the progress element", () => {
      const color = "secondary";

      const wrapper = mount(<ProgressBar color={color} />);

      expect(
        wrapper
          .find("ProgressBar")
          .find("progress")
          .hasClass(color)
      ).toEqual(true);
    });

    it("gives an error with invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<ProgressBar color="yellow" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("unfilled prop", () => {
    it("applies valid custom unfilled class to the progress element", () => {
      const unfilled = "dark";

      const wrapper = mount(<ProgressBar unfilled={unfilled} />);

      expect(
        wrapper
          .find("ProgressBar")
          .find("progress")
          .hasClass(unfilled)
      ).toEqual(true);
    });

    it("gives an error with invalid unfilled value", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<ProgressBar color="white" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("completion prop", () => {
    it("passes given completion value into the progress element", () => {
      const completion = 0.5;

      const wrapper = mount(<ProgressBar completion={completion} />);

      expect(
        wrapper
          .find("ProgressBar")
          .find("progress")
          .props().value
      ).toEqual(completion);
    });
  });
});
