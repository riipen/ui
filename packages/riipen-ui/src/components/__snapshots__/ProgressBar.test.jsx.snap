// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar> default props sets correct default props 1`] = `
<withClasses(ProgressBar)>
  <ProgressBar
    classes={
      Array [
        "riipen",
        "riipen-progressbar",
      ]
    }
    color="primary"
    completion={0}
    unfilled="light"
  >
    <div
      className="jsx-1920417447"
    >
      <progress
        className="jsx-1920417447 primary light riipen riipen-progressbar"
        value={0}
      />
      <JSXStyle
        dynamic={
          Array [
            "4px",
            "4px",
            "4px",
            "4px",
            "#1f71ae",
            "#1f71ae",
            "#fafafa",
            "#fafafa",
            "#dddddd",
            "#dddddd",
            "#f58632",
            "#f58632",
            "#3caabb",
            "#3caabb",
            "#4cc88d",
            "#4cc88d",
            "#e95353",
            "#e95353",
            "#febb46",
            "#febb46",
          ]
        }
        id="3168937156"
      />
    </div>
  </ProgressBar>
</withClasses(ProgressBar)>
`;
